x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+      return false;
+    }
+
+    this.logger.debug(`Lock "${this.lockId}" expired and was released.`);
+    return true;
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
+        },
+      },
+    });
+
+    const hits = result.hits.hits;
+    if (hits.length === 0) {
+      await this.releaseIfExpired();
+      return undefined;
+    }
+
+    return hits[0]._source;
+  }
+
+  public async waitForLock(options: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    const { metadata, ttlMs, ...pRetryOptions } = options;
+
+    return pRetry(async () => {
+      const acquired = await this.acquire({ metadata, ttlMs });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, pRetryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+
+  public async extendTtl(ttl = 300000): Promise<boolean> {
+    try {
+      await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        script: {
+          lang: 'painless',
+          source: `
+          long now = System.currentTimeMillis();
+          ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();`,
+          params: { ttl },
+        },
+        refresh: true,
+      });
+      this.logger.debug(`Lock "${this.lockId}" extended ttl with ${prettyMilliseconds(ttl)}.`);
+      return true;
+    } catch (error) {
+      if (
+        error instanceof errors.ResponseError &&
+        error.message.includes('version_conflict_engine_exception') &&
+        error.message.includes('but no document was found')
+      ) {
+        this.logger.debug(`Lock "${this.lockId}" was released concurrently. Not extending TTL.`);
+        return false;
+      }
+
+      this.logger.error(`Failed to extend lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+}
+
+export async function withLock<T>(
+  {
+    lockId,
+    esClient,
+    logger,
+    metadata,
+    ttlMs = 60 * 5 * 1000, // Default TTL: 5 minutes
+  }: {
+    lockId: LockId;
+    esClient: Client;
+    logger: Logger;
+  } & AcquireOptions,
+  callback: () => Promise<T>
+): Promise<T | undefined> {
+  const lockManager = new LockManager(lockId, esClient, logger);
+  const acquired = await lockManager.acquire({ metadata, ttlMs });
+
+  // extend the ttl periodically
+  const extendInterval = Math.floor(ttlMs / 2);
+  const intervalId = setInterval(async () => {
```

Comment:
Not a huge fan of passing an async function to setInterval etc, it can lead to uncaught promise rejections. I understand you're catching the error below, but it just doesn't seem like something we need. In fact you can do something like:

```
lockManager.extendTtl().catch(err => logger.error(`Failed to extend lock "${lockId}":`, err));
```